scripts/notebooks/SampleQC.ipynb

```diff
@@ -0,0 +1,2849 @@
+    "sex_ped_counts = pd.DataFrame.from_dict({'ID': list(updated_sex.keys()), 'SEX': list(updated_sex.values())})['SEX'].value_counts()\n",
+    "sex_ped_counts_normalized = pd.DataFrame.from_dict({'ID': list(updated_sex.keys()), 'SEX': list(updated_sex.values())})['SEX'].value_counts(normalize=True)\n",
+    "\n",
+    "# Display sex counts and frequencies\n",
+    "print(f\"Count of each unique value in column: {sex_ped_counts}\\n\")\n",
+    "print(f\"Proportion of each unique value in column: {sex_ped_counts_normalized}\\n\")\n",
+    "\n",
+    "plot_sex_chromosome_ploidy(samples_qc_table, updated_sex, display=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<div class=\"alert alert-block alert-success\"><b>If available</b>: If you have a PED file for your cohort, provide the Google Cloud Storage path to it via <tt>REFERENCE_ASSIGNMENTS_PED</tt>.</div>"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "REFERENCE_ASSIGNMENTS_PED = None\n",
+    "\n",
+    "if REFERENCE_ASSIGNMENTS_PED:\n",
+    "    validate_ped(REFERENCE_ASSIGNMENTS_PED, set(samples_qc_table['sample_id']))\n",
+    "else:  \n",
+    "    print(\"Reference PED file is not provided - skipping this step.\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "reference_ped = pd.DataFrame()\n",
+    "\n",
+    "if REFERENCE_ASSIGNMENTS_PED:\n",
+    "    reference_ped = process_reference_ped(REFERENCE_ASSIGNMENTS_PED, samples_qc_table)\n",
+    "else:\n",
+    "    print(\"Reference PED file is not provided - skipping this step.\")\n",
+    "    \n",
+    "reference_ped"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Create and display differences files using reference and computed sex assignments\n",
+    "differences_ped = pd.DataFrame()\n",
+    "\n",
+    "if REFERENCE_ASSIGNMENTS_PED:\n",
+    "    file_path = generate_file_path(TLD_PATH, 'sex_analysis', \"differing_sex_assignments.tsv\")\n",
+    "    differences_ped = create_ped_differences(reference_ped, updated_sex, file_path)\n",
+    "else:\n",
+    "    print(\"Reference PED file is not provided - skipping this step.\")\n",
+    "    \n",
+    "differences_ped"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Save and display generated PED file, and update 'cohort_ped' workspace variable to this\n",
+    "file_path = generate_file_path(TLD_PATH, 'sex_analysis', 'sample_qc.ped')\n",
+    "create_ped(updated_sex, reference_ped, file_path)\n",
+    "\n",
+    "created_ped = pd.read_table(file_path, names=['Family_ID', 'Sample_ID', 'Paternal_ID', \n",
+    "                                               'Maternal_ID', 'Sex', 'Phenotype'])\n",
```

Comment:
I'm getting this error:
```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Cell In[84], line 3
      1 # Save and display generated PED file, and update 'cohort_ped' workspace variable to this
      2 file_path = generate_file_path(TLD_PATH, 'sex_analysis', 'sample_qc.ped')
----> 3 create_ped(updated_sex, reference_ped, file_path)
      5 created_ped = pd.read_table(file_path, names=['Family_ID', 'Sample_ID', 'Paternal_ID', 
      6                                                'Maternal_ID', 'Sex', 'Phenotype'])
      7 created_ped

Cell In[30], line 25, in create_ped(sex_assignments, reference_ped, file_path)
     23 # Account for case where reference PED file does not exist
     24 if (reference_ped.empty):
---> 25     reference_ped["Family_ID"] = reference_ped["Sample_ID"]
     26     reference_ped = pd.DataFrame({'Sample_ID': sex_assignments.keys()})
     27     reference_ped["Paternal_ID"] = "0"

File /opt/conda/lib/python3.10/site-packages/pandas/core/frame.py:3761, in DataFrame.__getitem__(self, key)
   3759 if self.columns.nlevels > 1:
   3760     return self._getitem_multilevel(key)
-> 3761 indexer = self.columns.get_loc(key)
   3762 if is_integer(indexer):
   3763     indexer = [indexer]

File /opt/conda/lib/python3.10/site-packages/pandas/core/indexes/range.py:349, in RangeIndex.get_loc(self, key)
    347         raise KeyError(key) from err
    348 if isinstance(key, Hashable):
--> 349     raise KeyError(key)
    350 self._check_indexing_error(key)
    351 raise KeyError(key)

KeyError: 'Sample_ID'
```
which looks like a bug. I think this line just needs to be moved below the next one.

Reply:
Apologies, I'd noticed this post-hoc so updated it only in the template workspace rather than in the Github repo as well. I've included this version in the latest commit though.